### ISSUE TITLE
 Implement Leader Election for HA Mode

### ISSUE THREAD

**Issue by aminmr** [gianlucam76/k8s-cleaner#135]
## Description
Currently, the k8s-cleaner operator is deployed with a single replica by default and does not support high availability (HA). If the number of replicas is increased, multiple pods may attempt to take actions simultaneously, which could lead to conflicts or redundant operations on the cluster.

To address this, I suggest implementing leader election for HA mode. Many open-source projects utilize Kubernetes' Lease mechanism for this purpose, allowing only one pod to act as the leader at any given time. This would prevent multiple instances from interfering with each other when multiple replicas are running.

Here are the relevant Kubernetes docs: [Kubernetes Lease Mechanism](https://kubernetes.io/docs/concepts/architecture/leases/).

## Proposed Solution:

Introduce leader election logic using Kubernetes Leases.
Ensure that only the pod holding the active lease performs actions on the cluster, while other replicas remain in a standby mode.

I am happy to volunteer to implement this feature for the project.

I am looking forward to your thoughts! Thanks!

**Comment by gianlucam76** [comment]
Thanks @aminmr. I am not sure about this. k8s-cleaner originally had leader election. But I removed it. Reason being, I believe with cleaner we might hit scaling more than availability (and the end it does not have to respond to other services). And when we hit scaling, my plan is to introduce sharding. So different cleaner instances can process in parallel different cleaner instances based on some annotation.

Let's keep this open though. If we don't take that path, we will add leader election back. 

Thank you again!

**Comment by aminmr** [comment]
Thanks, @gianlucam76 !

I appreciate your explanation, but I have a couple of questions regarding the future implementation and design decisions for the Cleaner.
qir
**What technologies are you planning to use for implementing sharding in the Cleaner?**
I'm curious about how you plan to manage sharding to ensure multiple Cleaner instances can process resources in parallel.

**Could you clarify why leader election isn't a good solution for this project?**
I understand your point about scaling being more of a concern than availability, but I’m still unclear on why leader election is being considered less effective in this case. Is there a specific issue with leader election that conflicts with the overall goals of the project? And do you know any relevant operator with sharding feature?

Thanks again for your time, and I look forward to your thoughts on this!

**Comment by gianlucam76** [comment]
Hi @aminmr regarding sharing, I am planning on using same approach I used in [sveltos](https://projectsveltos.github.io/sveltos/scalability/sharding/) here it will require some manual configuration (as I don't have a shard controller like in Sveltos), but the idea is that.

In general leader election is great (tough it has a cost of you having to run 3 pods instead of 1 with 2 pods doing nothing most of the time). But I do see that more valuable for a service that need to respond to other services (where you cannot afford having it down for 30 seconds or so). 
Cleaner has a configurable [jitter window](https://github.com/gianlucam76/k8s-cleaner/blob/4ec9043198c3a3f7ebd971d7f9449f7a17078e3b/internal/controller/cleaner_controller.go#L291). So if cleaner gets stuck, it won't miss processing cleaner instances which are due. This makes the leader less needed.

**Comment by aaronmhmr** [comment]
Hi @gianlucam76 👋

Thanks again for taking the time to explain your reasoning — the sharding direction and the jitter-window argument both make sense to me, and I appreciate you leaving the issue open.

I've spent some time doing R&D on this: digging into the code (`cleaner_controller.go`, the manager setup, the chart) and working through what actually happens in the failure scenario that led me to open this issue — a node going unreachable, which I've hit in production.

I'd like to share what I found, because I think there's an important distinction between what **sharding** solves and what **leader election** solves, and I'm not convinced the two are substitutes. I'll lay out the scenario and a side-by-side comparison in the next comment so it's easy to follow.

I'd really appreciate it if you could briefly read through my R&D on this and share your opinion. **Whatever we land on, I'm happy to do the implementation work** — sharding, leader election, or both.

**Comment by aaronmhmr** [comment]
### The scenario: a node becomes unreachable

When a node's kubelet stops responding, its pods get a `deletionTimestamp` and enter `Terminating` — and stay there indefinitely, since only a kubelet can confirm a pod is really gone. The remedy is to force-delete them, which is exactly what I use k8s-cleaner for.

The problem: **if the k8s-cleaner pod is itself on that node, the Cleaner that would force-delete the stuck pods is itself a stuck pod.**

With the chart defaults (`replicaCount: 1`, no tolerations), recovery takes roughly **6 minutes**: ~40s for the node to be marked NotReady and tainted `node.kubernetes.io/unreachable:NoExecute`, then **300s** while the pod tolerates that taint (`--default-unreachable-toleration-seconds`, injected by the `DefaultTolerationSeconds` admission plugin, since the chart sets no tolerations).

To be precise: once the pod is finally evicted, the ReplicaSet **does** schedule a replacement on a healthy node, and it catches up. So this isn't a permanent outage — it's a **~6-minute hole during which k8s-cleaner is silently dead while the API still reports the pod `Running`.**

### Where the jitter window helps — and where it doesn't

You're right that the [jitter window](https://github.com/gianlucam76/k8s-cleaner/blob/4ec9043198c3a3f7ebd971d7f9449f7a17078e3b/internal/controller/cleaner_controller.go#L291) means we don't *miss* a run: `shouldSchedule()` processes any Cleaner whose `NextScheduleTime` has passed as soon as the new pod comes up. Correctness is fine.

But what this use case needs is **timeliness**, not eventual catch-up. A Cleaner that force-deletes stuck pods isn't much use if it fires six minutes after the outage it exists to remediate — the outage *is* those six minutes. The jitter window guarantees the run **happens**; it doesn't guarantee it happens **while it still matters**.

### Why sharding can't close that gap

Under sharding, each Cleaner has exactly one deployment permitted to process it, and each shard deployment is still `replicas: 1`. If the node hosting shard-N dies, shard-N's Cleaners are down for the same ~6 minutes and **no other shard may pick them up** — the shard predicate rejects them, by design.

So sharding doesn't remove the single point of failure; it gives me N of them and partitions my Cleaners across them. Whether my force-delete Cleaner survives a node loss then comes down to which shard I happened to annotate it into — luck, not design.

This isn't a criticism of sharding. It solves **throughput**, which is a real and separate problem. It just can't solve this one.

### Why a Lease can

With `replicas: 3` and anti-affinity, when the leader's node dies it stops renewing its Lease, and a standby acquires it in **~15–20s** (controller-runtime defaults: `LeaseDuration 15s`, `RenewDeadline 10s`, `RetryPeriod 2s`) and reconciles everything. Standbys already hold warm caches — controller-runtime starts `runnables.Caches` *before* leader election and gates only the controllers on the Lease — so there's no cold-start relist.

The property that I think decides this whole question:

> **Lease-based failover has no dependency on the dead pod being reaped.**

The old leader stays wedged in `Terminating` forever and it simply doesn't matter — the taint, the eviction, `tolerationSeconds` are all off the recovery path. Whereas the single-replica model's recovery is *entirely gated* on that eviction timeline, which is exactly the thing that breaks when a kubelet dies.

~20 seconds instead of ~6 minutes. And the new leader then force-deletes the stuck pods on the dead node — including its own dead predecessor.

| | **Sharding** | **Leader election** |
|---|---|---|
| Recovery | ~6 min (for the dead shard) | **~15–25 s** |
| Gated on the stuck `Terminating` pod? | Yes — via the eviction timeline | **No** |
| Failover for a given Cleaner | None — exactly one pod may run it | Any replica takes over |
| Solves | **Throughput** | **Availability** |

(Tuning `tolerationSeconds` to ~30s does help the single-replica case — ~6 min → ~1.5 min — and I use it today. But it's a faster reschedule, not a failover.)

### One more angle: k8s-cleaner as a fencing agent

Kubernetes does have a first-class answer to stuck `Terminating` pods — **Non-Graceful Node Shutdown** ([KEP-2268](https://github.com/kubernetes/enhancements/issues/2268)), **GA in v1.28**. Instead of force-deleting pods you taint the dead node:

```bash
kubectl taint nodes <node> node.kubernetes.io/out-of-service=nodeshutdown:NoExecute
```

PodGC then force-deletes the pods *and* their volumes get force-detached — control-plane side, no kubelet needed. Strictly better than a raw force-delete, which leaves volumes attached.

**But Kubernetes deliberately will not apply that taint for you.** The [docs](https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/) are explicit that you must first confirm the node is genuinely powered off, not merely unreachable — that's a **fencing** decision, and a wrong call means two pods writing the same volume. So it's punted to the operator, permanently and on purpose. Today the automation options are [kube-fencing](https://github.com/kvaps/kube-fencing) (dormant, last commit Feb 2025) or [medik8s](https://github.com/medik8s) (actively maintained, but a heavyweight OpenShift-oriented stack built around BMC/IPMI fence agents).

That leaves a real gap — and the rule that fills it is small and declarative: *"node NotReady for more than N minutes → apply the out-of-service taint."* **That is exactly the shape of a `Cleaner`**, and it needs no change to k8s-cleaner at all; a `Transform` action can set that taint today.

**Which is the point I'd like you to weigh.** The moment someone uses k8s-cleaner that way, **k8s-cleaner is a fencing agent** — and a fencing agent has one non-negotiable requirement: *it must be alive during the very node failure it exists to remediate.* Every serious fencing implementation — medik8s, kube-fencing, Pacemaker/STONITH before them — runs multiple replicas with leader election for exactly this reason.

I'm not claiming fencing is k8s-cleaner's primary use case. But given the actions a Cleaner can already take — deleting resources, transforming them, tainting Nodes — it's a natural one, and users will find it. In that role **HA stops being a nice-to-have and becomes mandatory**, and sharding cannot provide it.

### Proposal: they're orthogonal, and HA can be free

Sharding decides **what** each pod processes; leader election decides **who** processes it when a pod dies. They compose — Sveltos runs both. Here that's one Deployment per shard, each able to run `replicas: N` with its **own** Lease (`LeaderElectionID: k8s-cleaner-<shard-key>`; a single shared Lease ID would be a bug — all shards would contend and only one would run).

On your cost concern — *"3 pods instead of 1, with 2 doing nothing most of the time"* — that's fair, and I think it's avoidable. Ship leader election **opt-in and off by default**:

```yaml
replicaCount: 1        # unchanged
leaderElection:
  enabled: false       # unchanged behaviour for everyone
```

Then the default user pays nothing: same single pod, same footprint, no Lease. Only operators who explicitly need HA set `replicaCount: 3` and enable the flag — knowingly opting into that cost.

### What I'd suggest

Happy to implement both, as two separate PRs so you can take them independently and sharding stays your priority:

1. **Sharding** — a `--shard-key` flag plus the `sharding.projectsveltos.io/key` annotation, reusing `IsShardAMatch` from `libsveltos/lib/sharding` (already in our `go.mod`), plus per-shard deployments in the chart.
2. **Leader election** — `--leader-elect` (default `false`), with a per-shard `LeaderElectionID`.

Does this seem reasonable? Happy to adjust if you see it differently — and happy to start on the sharding PR right away regardless.
